third_party/move/move-compiler-v2/src/pipeline/control_flow_graph_simplifier.rs

```diff
@@ -0,0 +1,687 @@
+        Self::replace_blocks(self.successors_mut(block), old_succ, new_succ);
+    }
+
+    /// Replaces `old` by `new` in `blocks`
+    fn replace_blocks(blocks: &mut [BlockId], old: BlockId, new: BlockId) {
+        for block in blocks {
+            if *block == old {
+                *block = new;
+            }
+        }
+    }
+
+    /// Checks if the control flow graph is consistent
+    /// - `self.code_blocks`, `self.successors`, and `self.predecessors` have the same set of keys
+    /// - `self.successors` is the reverse of `self.predecessors`
+    /// - TODO: check if the code blocks are consistent with the successors and predecessors map
+    fn check_consistency(&self) {
+        assert!(self.code_blocks.keys().collect_vec() == self.successors.keys().collect_vec());
+        assert!(self.code_blocks.keys().collect_vec() == self.predecessors.keys().collect_vec());
+        let pred_map_reversed = self.predecessors.iter().fold(
+            BTreeMap::new(),
+            |mut acc: BTreeMap<BlockId, Vec<BlockId>>, (block, preds)| {
+                acc.entry(*block).or_default();
+                for pred in preds {
+                    acc.entry(*pred).or_default().push(*block);
+                }
+                acc
+            },
+        );
+        assert!(self.successors.keys().collect_vec() == pred_map_reversed.keys().collect_vec());
+        for block in self.successors.keys() {
+            assert!(
+                self.successors
+                    .get(block)
+                    .expect("successors")
+                    .iter()
+                    .collect::<BTreeSet<_>>()
+                    == pred_map_reversed
+                        .get(block)
+                        .expect("pred_map_reversed")
+                        .iter()
+                        .collect::<BTreeSet<_>>()
+            );
+        }
+    }
+}
+
+/// Transformation state for transformation 1 in the module doc
+struct EmptyBlockRemover(ControlFlowGraphCodeGenerator);
+
+impl EmptyBlockRemover {
+    /// Performs the transformation 1 described in the module doc
+    pub fn transform(generator: ControlFlowGraphCodeGenerator) -> ControlFlowGraphCodeGenerator {
+        let mut processer = Self::new(generator);
+        processer.process();
+        processer.0
+    }
+
+    /// Wrapper
+    fn new(generator: ControlFlowGraphCodeGenerator) -> Self {
+        Self(generator)
+    }
+
+    /// Performs the transformation 1 described in the module doc
+    fn process(&mut self) {
+        for block in self.0.blocks() {
+            if self.is_empty_block(block) {
+                let succ_block = self.0.get_the_non_trivial_successor(block);
+                if succ_block != block {
+                    self.remove_empty_block(block, succ_block);
+                }
+            }
+        }
+    }
+
+    /// Checks if the given block is empty; i.e., only a label and a jump
```

Comment:
Clearly, not for "all" predecessors. Why not for entry block? Would it still be okay if we had something like:

```
start
L1: goto L2
```

Reply:
The entry block is purely formal. The "real" entry block is actually the successor of the formal entry block, which we  do process.